### PR TITLE
fix: WebAssembly Compilation

### PR DESCRIPTION
**Pull request by troyedwardsjr**
Fixed several type errors when compiling to WebAssembly, added flags to `build.rs`, and added instructions to compile to WebAssembly.

For example:
```
error[E0308]: mismatched types
   --> src/component.rs:118:9
    |
118 |         size: desc.layout.size() as u64,
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u32`, found `u64`

error[E0308]: mismatched types
   --> src/component.rs:119:14
    |
119 |         alignment: desc.layout.align() as u64,
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u32`, found `u64`

```

`systems.rs` example Chrome logs:
image

### CODE REVIEW

File: src/component.rs [hunk 1]
```diff
@@ -113,12 +113,19 @@ pub fn register_component(world: *mut ecs_world_t, desc: ComponentDescriptor) ->
 	
 	// let s_id = 0;
 	let comp_desc = ecs_component_desc_t {
-        _canary: 0,
+		_canary: 0,
 		entity: e_desc,
+		#[cfg(all(target_arch = "wasm32", target_os = "emscripten"))]
```
Comment (jazzay) [inline src/component.rs]: I wonder if we can replace all these with size_t instead, and avoid the cfg checks?

File: src/system.rs [hunk 2]
```diff
@@ -280,58 +280,65 @@ impl Iter {
...
+				* single value. Ensure that the application does not accidentally read
+				* out of bounds. */
+			if is_shared {
+					count = 1;
+			}
+			// println!("Term: {}, is_shared: {}, count: {}", term_id, is_shared, count);
+
+			let size = std::mem::size_of::<T>();
+			#[cfg(all(target_arch = "wasm32", target_os = "emscripten"))]
+			let array = unsafe { ecs_term_w_size(self.it, size as u32, index) as *mut T };
```
Comment (jazzay) [inline src/system.rs]: size_t might make this cleaner.

File: src/world.rs [hunk 3]
```diff
@@ -297,8 +297,12 @@ impl Drop for World {
 // Additional Add-ons support
 impl World {
 	pub fn enable_rest(&self) {
-		let rest_comp_id = unsafe { FLECS__EEcsRest as u64 };
+    let rest_comp_id = unsafe { FLECS__EEcsRest as u64 };
+		#[cfg(all(target_arch = "wasm32", target_os = "emscripten"))]
+		let rest_comp_size = std::mem::size_of::<EcsRest>() as u32;
```
Comment (jazzay) [inline src/world.rs]: size_t?

**Review by jazzay — COMMENTED** [review]
Looks great! 
Perhaps we can eliminate the cfg checks by using size_t instead. 

This makes me realize that the current setup for bindings.rs (writing into src folder) may not good for platform specific bindgen? The default is normally to write that to the target folder.

Have you tried building for native and wasm platforms?

**Comment by troyedwardsjr** [comment]
> Looks great! Perhaps we can eliminate the cfg checks by using size_t instead.

Thanks! Makes sense, updated.

> This makes me realize that the current setup for bindings.rs (writing into src folder) may not good for platform specific bindgen? The default is normally to write that to the target folder.

Yeah, I agree, it gets overwritten on every target switched build, which it doesn't seem to catch on some builds for whatever reason (perhaps it needs more `rerun-if-changed` calls), which causes errors on the other target.

> Have you tried building for native and wasm platforms?

Yup, works on both!

```shell
$ cargo run --example systems    
Finished dev [unoptimized + debuginfo] target(s) in 0.06s
Running `target/debug/examples/systems`
Systems example starting...
Sys1 - A-0: Position { x: 3.0, y: 6.0 
}, Velocity { x: 2.0, y: 4.0 }
...
```

```
$ cargo build --example systems --target wasm32-unknown-emscripten
Finished dev [unoptimized + debuginfo] target(s) in 0.07s
```

**Pull request merged**